go-controller/pkg/metrics/master.go

```diff
@@ -698,6 +728,520 @@ func getPodUIDFromPortBinding(row *sbdb.PortBinding) kapimtypes.UID {
+		UpdateFunc: func(table string, old model.Model, new model.Model) {
+			if table != nbGlobalTable {
+				return
+			}
+			oldRow := old.(*nbdb.NBGlobal)
+			newRow := new.(*nbdb.NBGlobal)
+
+			if oldRow.HvCfg != newRow.HvCfg && oldRow.HvCfgTimestamp != newRow.HvCfgTimestamp && newRow.HvCfgTimestamp > 0 {
+				select {
+				case updateOVNMeasurementCh <- hvCfgUpdate{hvCfg: newRow.HvCfg, timestamp: newRow.HvCfgTimestamp}:
+				default:
+					klog.Warning("Config duration recorder: unable to update OVN measurement")
+					select {
+					case deleteOVNMeasurementCh <- newRow.HvCfg:
+					default:
+					}
+				}
+			}
+		},
+	})
+	cr.enabled = true
+}
+
+// Start allows the caller to attempt measurement of a control plane configuration duration, as a metric,
+// the duration between functions Start and End. Optionally, if you wish to record OVN config duration,
+// call AddOVN which will add the duration for OVN to apply the configuration to all nodes.
+// The caller must pass kind,namespace,name and object resource version which will be used to determine if the object
+// is allowed to record. To allow no locking, each go routine that calls this function, can determine itself
+// if it is allowed to measure.
+// There is a mandatory two-step process to complete a measurement.
+// Step 1) Call Start when you wish to begin a measurement - ideally when processing for the object starts
+// Step 2) Call End which will complete a measurement
+// Optionally, call AddOVN when you are making a transaction to OVN in order to add on the OVN duration to an existing
+// measurement. This must be called between Start and End. Not every call to Start will result in a measurement
+// and the rate of measurements depends on the number of nodes and function Run arg k.
+// Only one measurement for a kind/namespace/name is allowed until the current measurement is Ended (via End) and
+// processed. This is guaranteed by workqueues (even with multiple workers) and event handlers.
+func (cr *ConfigDurationRecorder) Start(kind, namespace, name string) (time.Time, bool) {
+	if !cr.enabled {
+		return time.Time{}, false
+	}
+	kindNamespaceName := fmt.Sprintf("%s/%s/%s", kind, namespace, name)
+	if !cr.allowedToMeasure(kindNamespaceName) {
+		return time.Time{}, false
+	}
+	measurementTimestamp := time.Now()
+	cr.measurementsMu.Lock()
+	_, found := cr.measurements[kindNamespaceName]
+	// we only record for measurements that aren't in-progress
+	if !found {
+		cr.measurements[kindNamespaceName] = measurement{kind: kind, startTimestamp: measurementTimestamp,
+			start: true, expiresAt: measurementTimestamp.Add(maxMeasurementLifetime)}
+	}
+	cr.measurementsMu.Unlock()
+	return measurementTimestamp, !found
+}
+
+// allowedToMeasure determines if we are allowed to measure or not. To avoid the cost of synchronisation by using locks,
+// we use probability.
+func (cr *ConfigDurationRecorder) allowedToMeasure(kindNamespaceName string) bool {
```

Comment:
add comment about `allowedToMeasure` requiring to return the same value for the same `kindNamespaceName`